dask/dataframe/io/parquet/arrow.py

```diff
@@ -625,6 +654,181 @@ def _construct_parts(
 
 
+#
+#  PyArrow Dataset API [PyArrow>=1.0.0] - Private Helper Functions
+#
+
+
+def _get_all_partition_keys(ds):
+    pkeys = defaultdict(list)
+    for file_frag in ds.get_fragments():
+        keys = pa_ds._get_partition_keys(file_frag.partition_expression)
+        for k, v in keys.items():
+            if v not in pkeys[k]:
+                pkeys[k].append(v)
+    return pkeys
+
+
+def _collect_pyarrow_dataset_frags(ds, filters):
+    partition_names = []
+    partition_obj = []
+    partition_keys = {}
+    metadata = []
+    ds_filters = None
+    if filters is not None:
+        ds_filters = pq._filters_to_expression(filters)
+    pkeys = None
+    for file_frag in ds.get_fragments(ds_filters):
+        for rg_frag in file_frag.split_by_row_group(ds_filters, schema=ds.schema):
+            metadata.append(rg_frag)
+            keys = pa_ds._get_partition_keys(rg_frag.partition_expression)
+            if keys:
+                if pkeys is None:
+                    # Get all partition keys without filters...
+                    pkeys = _get_all_partition_keys(ds)
+                partition_keys[rg_frag.path] = list(keys.items())
+                if not partition_names:
+                    partition_names = list(keys)
+
+    for name in partition_names:
+        partition_obj.append(PartitionObj(name, pkeys[name]))
+
+    metadata = sorted(metadata, key=lambda x: natural_sort_key(x.path))
+    partition_info = {
+        "partitions": partition_obj,
+        "partition_keys": partition_keys,
+        "partition_names": partition_names,
+    }
+    return metadata, partition_info
+
+
+def _gather_metadata_pyarrow_dataset(
+    paths, fs, split_row_groups, gather_statistics, filters, dataset_kwargs
+):
+    """ pyarrow.dataset version of _gather_metadata
+        Use pyarrow.dataset API to collect list of row-group fragments.
+        Also, collect other information necessary for parquet-to-ddf
+        mapping (e.g. schema, partition_info).
+    """
+    # Use pyarrow.dataset API
+    ds = None
+    if len(paths) == 1 and fs.isdir(paths[0]):
+
+        # Use _analyze_paths to avoid relative-path
+        # problems (see GH#5608)
+        base, fns = _analyze_paths(paths, fs)
+        paths = fs.sep.join([base, fns[0]])
+
+        meta_path = fs.sep.join([paths, "_metadata"])
+        if fs.exists(meta_path):
+            # Use _metadata file if available
+            ds = pa_ds.parquet_dataset(
+                meta_path, partitioning=dataset_kwargs.get("partitioning", "hive")
+            )
+            if gather_statistics is None:
+                gather_statistics = True
+    elif len(paths) > 1:
+        base, fns = _analyze_paths(paths, fs)
+        meta_path = fs.sep.join([base, "_metadata"])
+        if "_metadata" in fns:
+            # Use _metadata file if available
+            # Pyarrow cannot handle "_metadata" when `paths` is a list
+            paths.remove(meta_path)
+            fns.remove("_metadata")
```

Comment:
What is actually the expected behaviour here?  Should what is defined in `_metadata` be filtered with the other specified paths? 

Because with this, it will ignore the file paths that are in `paths`? (as it will only look at the `_metadata` and construct the dataset from that)

Reply:
What I actually wanted to highlight (apart from the potentially unneeded "remove" operation): what is the intent of the user by passing a list of file paths? Because *if* there is a `_metadata` file in that list, the remaining content of the list is ignored, and you basically put anything in there.  
For example, if the user is assuming that listing a subset of files in that list would limit reading to those specified files, then that is actually not happening.

Reply:
> Because if there is a _metadata file in that list, the remaining content of the list is ignored, and you basically put anything in there.

Ah- that is a **very** good point.  Thanks for bringing attention to this!

In that case, it seems that we need to pass the file list along to `_collect_pyarrow_dataset_frags`, and ignore any file fragments that are not in the original path list.  Does that approach make sense?

Reply:
Yes, that approach seems to make sense! (although it's still a bit strange use case ;-))

Reply:
> although it's still a bit strange use case ;-)

Yes- It is certainly a special case.  However, I have had users insist that they need to pass in a list of files, and it would be nice to leverage a _metadata file if it is available.  I'll admit the number of people who care about this is likely small...